### ISSUE TITLE
Improve Windows building instructions

### ISSUE THREAD

**Issue by nh2** [facebook/rocksdb#3600]
https://github.com/facebook/rocksdb/wiki/Building-on-Windows says

> 5. wget

but it's not clear where to get it; I tried the Windows version from Sourceforge but it doesn't seem to find its libs on my Windows.

> All of the following is executed from the "Developer Command Prompt for VS2015":
> ...
> `unzip v2.2.0.zip`

`unzip` is not available on my Windows.

Or is this somehow assuming some specific options chosen in the installer from

>  Git - I used the Windows Git Bash.

?

Thanks!

**Comment by nh2** [comment]
Also what's not mentioned

```
> hg import exports.patch
applying exports.patch
abort: no username supplied
(use 'hg config --edit' to set your username)
```

which one can easily overlook when following the instructions.

**Comment by nh2** [comment]
Another thing that isn't clear is which subcomponents of

> Microsoft Visual Studio 2015 (Community)

have to be installed. When I installed the default selection, it first failed at gflags saying that some files related to some Windows 8.1 SDK are missing (fixed by running the installer again and manualy adding `Windows and Web development -> Windows 8.1 and Windows Phone 8.0/8.1 Tools -> Tools and Windows SDKs`), and later at snappy with `Project 'snappy' could not be loaded because it's missing install components. ... Install Windows XP Support for C++`  (which I fixed by selecting from the installer `Programming Languages -> Visual C++ -> Windows XP Support for C++`).

**Comment by nh2** [comment]
Then at cmake time in rocksdb, I get:

```
-- JNI library is enabled
-- Could NOT find Java (missing: Java_JAVA_EXECUTABLE Java_JAR_EXECUTABLE Java_JAVAC_EXECUTABLE Java_JAVAH_EXECUTABLE Java_JAVADOC_EXECUTABLE)
-- Could NOT find JNI (missing: JAVA_AWT_LIBRARY JAVA_JVM_LIBRARY JAVA_INCLUDE_PATH JAVA_INCLUDE_PATH2 JAVA_AWT_INCLUDE_PATH)
Downloading C:/Users/Administrator/code/rocksdb/java/test-libs/junit-4.12.jar
Downloading C:/Users/Administrator/code/rocksdb/java/test-libs/hamcrest-core-1.3.jar
Downloading C:/Users/Administrator/code/rocksdb/java/test-libs/mockito-all-1.10.19.jar
Downloading C:/Users/Administrator/code/rocksdb/java/test-libs/cglib-2.2.2.jar
Downloading C:/Users/Administrator/code/rocksdb/java/test-libs/assertj-core-1.7.1.jar
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
JAVA_AWT_INCLUDE_PATH (ADVANCED)
   used as include directory in directory C:/Users/Administrator/code/rocksdb/java
   used as include directory in directory C:/Users/Administrator/code/rocksdb/java
   used as include directory in directory C:/Users/Administrator/code/rocksdb/java
   used as include directory in directory C:/Users/Administrator/code/rocksdb/java
   used as include directory in directory C:/Users/Administrator/code/rocksdb/java
JAVA_INCLUDE_PATH (ADVANCED)
   used as include directory in directory C:/Users/Administrator/code/rocksdb/java
   used as include directory in directory C:/Users/Administrator/code/rocksdb/java
   used as include directory in directory C:/Users/Administrator/code/rocksdb/java
   used as include directory in directory C:/Users/Administrator/code/rocksdb/java
   used as include directory in directory C:/Users/Administrator/code/rocksdb/java
JAVA_INCLUDE_PATH2 (ADVANCED)
   used as include directory in directory C:/Users/Administrator/code/rocksdb/java
   used as include directory in directory C:/Users/Administrator/code/rocksdb/java
   used as include directory in directory C:/Users/Administrator/code/rocksdb/java
   used as include directory in directory C:/Users/Administrator/code/rocksdb/java
   used as include directory in directory C:/Users/Administrator/code/rocksdb/java

-- Configuring incomplete, errors occurred!
See also "C:/Users/Administrator/code/rocksdb/build/CMakeFiles/CMakeOutput.log".
See also "C:/Users/Administrator/code/rocksdb/build/CMakeFiles/CMakeError.log".
```

So probably there are some Java deps not mentioned in the `Pre-requisites`.

I removed `-DJNI=1` to work around that as I don't need java support.

Then I got

```
CMake Warning:
  Manually-specified variables were not used by the project:

    LZ4
```

so probably the `-DLZ4=1` given in the instructions is outdated?
```

**Comment by nh2** [comment]
Next, it failed at

```
C:\Users\Administrator\code\rocksdb\build>msbuild rocksdb.sln /p:Configuration=Release
Microsoft (R) Build Engine version 14.0.25420.1
Copyright (C) Microsoft Corporation. All rights reserved.

Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
Build started 3/13/2018 12:56:33 AM.
Project "C:\Users\Administrator\code\rocksdb\build\rocksdb.sln" on node 1 (default targets).
C:\Users\Administrator\code\rocksdb\build\rocksdb.sln.metaproj : error MSB4126: The specified solution configuration "R
elease|X86" is invalid. Please specify a valid solution configuration using the Configuration and Platform properties (
e.g. MSBuild.exe Solution.sln /p:Configuration=Debug /p:Platform="Any CPU") or leave those properties blank to use the
default solution configuration. [C:\Users\Administrator\code\rocksdb\build\rocksdb.sln]
Done Building Project "C:\Users\Administrator\code\rocksdb\build\rocksdb.sln" (default targets) -- FAILED.


Build FAILED.

"C:\Users\Administrator\code\rocksdb\build\rocksdb.sln" (default target) (1) ->
(ValidateSolutionConfiguration target) ->
  C:\Users\Administrator\code\rocksdb\build\rocksdb.sln.metaproj : error MSB4126: The specified solution configuration
"Release|X86" is invalid. Please specify a valid solution configuration using the Configuration and Platform properties
 (e.g. MSBuild.exe Solution.sln /p:Configuration=Debug /p:Platform="Any CPU") or leave those properties blank to use th
e default solution configuration. [C:\Users\Administrator\code\rocksdb\build\rocksdb.sln]

    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:00.09
```

I think that step should be

```
msbuild rocksdb.sln /p:Configuration=Release /p:Platform=x64
```

with `/p:Platform=x64` like for the other projects, instead of

```
msbuild rocksdb.sln /p:Configuration=Release
```

**Comment by rafalsk** [comment]
PLEASE DO UPDATE THE VERSION NUMBERS OF EXTERNAL LIBRARIES USED.

We've made sure, double checked that we followed the tutorial and still we are left with 37 missing dependencies (in terms of the number of functions) when linked with our software.

**Comment by rafalsk** [comment]
What makes me wonder; The codebase of snappy hasn't been updated since the last tutorial update and still we do get such an error when building on Windows:
```
"Severity	Code	Description	Project	File	Line	Suppression State
Error	LNK2019	unresolved external symbol "__declspec(dllimport) bool __cdecl snappy::RawUncompress(char const *,unsigned __int64,char *)" (__imp_?RawUncompress@snappy@@YA_NPEBD_KPEAD@Z) referenced in function "bool __cdecl rocksdb::Snappy_Uncompress(char const *,unsigned __int64,char *)" (?Snappy_Uncompress@rocksdb@@YA_NPEBD_KPEAD@Z)	GRIDNET	C:\Users\rafal\Documents\Visual Studio 2017\GRIDNET\GRIDNET\rocksdb.lib(format.cc.obj)	1	
"
```

**Comment by rafalsk** [comment]
When we ADDITIONALLY link with snappy64.lib the particular problem disappears. But I don't suppose that's the point of providing static libs in the rocksdb build process.

**Comment by gfosco** [comment]
You're right that this info was really old..  I've been going through this and creating a new version of that wiki page, using VS2017... I'm planning to remove Snappy and Zlib, instead using LZ4 and ZSTD for compression.  I'm certain it will be using different version numbers of dependencies, as a lot has changed.

**Comment by rafalsk** [comment]
@gfosco thats for a response but it seems the versions might be right after all; there's something wrong in how the build process included the depeandancies. Take Rpcrt4.lib for instance.

**Comment by rafalsk** [comment]
It's out of the scope of this problem; but what path convention are we to use on windows?
"C:\path\file" doesn't work

**Comment by gfosco** [comment]
cc @adamretter if you have thoughts here or an answer on the path issue

**Comment by siying** [comment]
@rafalsk by the way, you don't have to build all the compression libraries. You can skip those ones you don't plan to use. In fact, if you only plan to use XPRESS, you don't need to try out any of those compression libraries. Just set up the proper flag when running cmake and you'll be good to go.

**Comment by siying** [comment]
CC @yuslepukhin for the path question.

**Comment by rafalsk** [comment]
I'm just starting the journey with RocksDB, yes - I supposed those switches to cmake were there for a reason:) I was just trying to make it work without even looking up what those libraries were for.  I suppose one can choose the compression algorithm on the fly; but for this, I won't be taking your time and read docs instead.

**Comment by yuslepukhin** [comment]
@siying Sorry I was not aware that these instructions exists. Will review. I build rocksdb everyday (sans JNI) and it just works.
@rafalsk Please, clarify the context of your question about the path.

**Comment by yuslepukhin** [comment]
Here is how I do build rocksdb on my dev box.
 Get Git for Windows, CMake as described in the aforementioned instructions.
I have both 2015 and 2017 Studios installed. Here is the components that I have for 2017 C++.  

![vs 2017 components](https://user-images.githubusercontent.com/11303988/37741558-12ac7b44-2d1f-11e8-8978-08b8b8d7685f.PNG)

Here is the components for VS 2015
![2015_1](https://user-images.githubusercontent.com/11303988/37741867-63884e0c-2d20-11e8-9ece-50373405d703.PNG)
![2015_2](https://user-images.githubusercontent.com/11303988/37741879-718da3da-2d20-11e8-9110-c8fcfb49a951.PNG)

We currently target 10.0.10240 SDK so my build picks up 2015 SDK installation but feel free to install whatever SDK you want with 2017.

To unzip the zip files on windows simply use Explorer 
that treats it as folders and copy files out of that folder. Alternatively, right click and Extract All.

To unzip from PowerShell use something like this:
Add-Type -AssemblyName System.IO.Compression.FileSystem ; 
[System.IO.Compression.ZipFile]::ExtractToDirectory("C:\Users\dmitrism\Downloads\gflags.zip")

thirdparty.inc has some hard-coded paths which may be overridden by env variables. We use internal packages so I simply symlink %THIRD_PARTY% directory to package locations like this:
` Directory of D:\THIRD_PARTY

03/15/2018  08:11 PM    <DIR>          .
03/15/2018  08:11 PM    <DIR>          ..
02/01/2018  05:15 PM    <SYMLINKD>     Gflags.Library [D:\CorextCache\Gflags.1.0.0]
02/01/2018  05:15 PM    <SYMLINKD>     Snappy.Library [D:\CorextCache\Snappy.Library.1.1.0]
03/15/2018  08:11 PM    <SYMLINKD>     ZSTD.Library [D:\CorextCache\ZSTD.Library.1.3.2-2]`

Or you can override it with env vars like this:

`set SNAPPY_INCLUDE=%1\Snappy.Library\build\native\inc\inc
set SNAPPY_LIB_DEBUG=%1\Snappy.Library\lib\native\debug\amd64\snappy.lib
set SNAPPY_LIB_RELEASE=%1\Snappy.Library\lib\native\retail\amd64\snappy.lib

set ZSTD_INCLUDE=%1\ZSTD.Library\build\native\inc
set ZSTD_LIB_DEBUG=%1\ZSTD.Library\lib\native\debug\amd64\libzstd_static.lib
set ZSTD_LIB_RELEASE=%1\ZSTD.Library\lib\native\retail\amd64\libzstd_static.lib`

Note that XPRESS is a built-in Windows compression and that does not require any dependency although not as good as ZSTD.

So after you cloned rocksdb into a folder and providing all the dependencies paths are setup I do the following:

1. `md build && cd build`
2. `cmake -G "Visual Studio 15 Win64" -DCMAKE_SYSTEM_VERSION=10.0.10240 -DWITH_GFLAGS=1 -DWITH_SNAPPY=1 -DWITH_JEMALLOC=1 -DWITH_XPRESS  ..`
3. Win64 generator ensures that the solution is x64 and not x86. You can specify the SDK you want to build with. Feel free to add more dependancies as you see fit.
3. To build debug simply type `msbuild rocksdb.sln /m`. Debug is default and architecture is x64 by default since this is what you specified with CMake. Note that this is a cached CMake property. So if you simply ran CMake first without Win64 it will be x86 even after you rerun it. You need to blast the build directory and rerun CMake.
4. To build release `msbuild rocksdb.sln /m /p:Configuration=Release`

So this is it. You should have it. 

Now if you want to run tests you can do so with `run_ci_db_test.ps1` script which runs things in parallel. I usually run against a RAM drive. Download imdisk and install, it is quick but not silent and for that reason I could not do it on AppVeyor.

To create a RAM drive on F: with 3Gb of space run something like this:
imdisk -a -s 3000M -m  f:  -p "/fs:ntfs /q /y"

Then open Powershelll prompt, cd to your rocksdb source tree and run something like this. -EnableJE is needed if you built with Jemalloc.

`.\build_tools\run_ci_db_test.ps1 -EnableJE -Concurrency 6 -WorkFolder f:\db -ExcludeExes c_test,backupable_db_test -RunAll`

**Comment by trustfarm-dev** [comment]
https://github.com/trustfarm-dev/rocksdb-2.7-win64-buildrelease

I've also several tweak and helps from resources in internet,
built on MSVS2017 x64 latest full binaries and libraries.
It may be helpful to don't want to spend time to make rocksdb windows stuffs.

**Comment by ramvadiv** [comment]
Closing this old issue. The Windows build page (https://github.com/facebook/rocksdb/wiki/Building-on-Windows ) has gone thru significant updates after this issue was created including step-by-step instructions. Pls open a new issue if you still have problems with Windows build.
